src/categorical_algebra/DPO.jl

```diff
@@ -0,0 +1,215 @@
+module DPO
+export rewrite, rewrite_match, valid_dpo, dangling_condition, id_condition, pushout_complement, extend_cset,subcset
+
+using ..FinSets
+import ..CSets: AbstractACSet, ACSet, ACSetTransformation, components, homomorphism, homomorphisms, unpack_components, pack_components, add_parts!, nparts, set_subpart!, is_natural
+using ...Theories
+using ..Limits
+
+"""
+    l
+  L ← I
+m ↓   ↓k
+  G ← K
+    g
+
+Compute a pushout complement, componentwise in Set. On a formal level:
+For each component, define K = G / m(L/l(I)). There is a natural injection g: K↪G
+For each component, define k as equal to the map l;m (every element in the image in G is also in K).
+
+Returns ACSetTransformations k and g such that (m, g) is the pushout of (l, k)
+
+Implementation-wise, elements of K are ordered in the same order as they appear in G.
+Construct an offset mapping to keep track of the location of elements of G within K.
+"""
+function pushout_complement(l::ACSetTransformation,m::ACSetTransformation)::Pair{ACSetTransformation,ACSetTransformation}
+
+  I, L, G = l.dom, l.codom, m.codom
+  lm = compose(l, m)
+
+  k_components = Dict{Symbol, FinFunction}()
+  g_components = Dict{Symbol,Vector{Int}}()
+  K = ACSet{typeof(G).parameters...}()
+  offsets = Dict{Symbol,Vector{Int}}()
+
+
+  for comp in keys(l.components)
+
+    n_comp = nparts(G, comp)  # total # of elements in G
+    l_image = Set(l.components[comp].func)
+
+    # m(L/l(G)): image of (complement of image of l) into G
+    orphans = sort(map(x->m[comp](x),
+                  filter(x->!(x in l_image),
+                      1:nparts(L,comp))))
+    orph_set = Set(orphans)  # for membership test
+
+    # Compute map from G to K (undefined for orphans)
+    offsets[comp] = reindex(n_comp, orphans)
+
+    # Natural injection of subset of G into G
+    g_components[comp] = filter(x->!(x in orph_set), 1:n_comp)
+
+    # Initialize component in K
+    add_parts!(K, comp, length(g_components[comp]))
+
+    # Adjust lg function from I→G to refer to K using the offset
+    newFunc = [x - offsets[comp][x] for x in lm.components[comp].func]
+    k_components[comp] = FinFunction(newFunc, n_comp - length(orphans))
+  end
+
+  # Populate attributes for K (unchanged from G)
+  comps, arrows, src, tgt = typeof(G).parameters[1].parameters
+  attrs, srcs = typeof(G).parameters[2].parameters[3:4]
```

Comment:
Unpacking parameters like this unreadable and error-prone. Use the accessors in `theories/Schema.jl` for this.